src/imperator/data/events.rs

```diff
@@ -0,0 +1,252 @@
+            if let Ok(id) = u16::from_str(id) {
+                return self.events.get(&(namespace.to_string(), id));
+            }
+        }
+        None
+    }
+
+    pub fn check_scope(&self, token: &Token, sc: &mut ScopeContext) {
+        if let Some(event) = self.get_event(token.as_str()) {
+            sc.expect(event.expects_scope, &Reason::Token(token.clone()));
+        }
+    }
+
+    pub fn namespace_exists(&self, key: &str) -> bool {
+        self.namespaces.contains_key(key)
+    }
+
+    pub fn iter_namespace_keys(&self) -> impl Iterator<Item = &Token> {
+        self.namespaces.values()
+    }
+
+    pub fn exists(&self, key: &str) -> bool {
+        if let Some((namespace, id)) = key.split_once('.') {
+            if let Ok(id) = u16::from_str(id) {
+                if self.events.contains_key(&(namespace.to_string(), id)) {
+                    return true;
+                }
+            }
+        }
+        false
+    }
+
+    pub fn iter_keys(&self) -> impl Iterator<Item = &Token> {
+        self.events.values().map(|item| &item.key)
+    }
+
+    pub fn validate(&self, data: &Everything) {
+        for item in self.events.values() {
+            item.validate(data);
+        }
+    }
+}
+
+impl FileHandler<Block> for ImperatorEvents {
+    fn subpath(&self) -> PathBuf {
+        PathBuf::from("events")
+    }
+
+    fn load_file(&self, entry: &FileEntry) -> Option<Block> {
+        if !entry.filename().to_string_lossy().ends_with(".txt") {
+            return None;
+        }
+
+        PdxFile::read(entry)
+    }
+
+    fn handle_file(&mut self, _entry: &FileEntry, mut block: Block) {
+        for item in block.drain() {
+            if let Some(Field(key, _, bv)) = item.expect_into_field() {
+                if key.is("namespace") {
+                    if let Some(value) = bv.expect_into_value() {
+                        self.namespaces.insert(value.to_string(), value);
+                    }
+                } else if let Some(block) = bv.into_block() {
+                    self.load_event(key, block);
+                } else {
+                    let msg = "unknown setting in event files";
+                    error(key, ErrorKey::UnknownField, msg);
+                }
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Event {
+    key: Token,
+    block: Block,
+    expects_scope: Scopes,
+    expects_from_token: Token,
+}
+
+const EVENT_TYPES: &[&str] = &[
+    "character_event",
+    "minor_character_event",
+    "country_event",
+    "minor_country_event",
+    "major_country_event",
+    "state_event",
+    "province_event"
+];
+
+impl Event {
+    pub fn new(key: Token, block: Block) -> Self {
+        let mut expects_scope = Scopes::Country;
+        let mut expects_from_token = key.clone();
+
+        if let Some(event_type) = block.get_field_value("type") {
+            match event_type.as_str() {
+                "character_event" => {
```

Comment:
Clippy will tell you to combine the arms with identical bodies, like `"character_event" | "minor_character_event" => `

Up to you whether to do that or add `#[allow(clippy::match_same_arms)]` here, it's mostly a style issue.